lib/async_hooks.js

```diff
@@ -203,6 +205,141 @@ class AsyncResource {
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      for (let i = 0; i < locals.length; ++i) {
+        locals[i]._propagate(resource, triggerResource);
+      }
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      for (let i = 0; i < locals.length; ++i) {
+        delete resource[locals[i].kResourceStore];
+      }
+      idResourcePairing.delete(asyncId);
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
+    asyncContextHook.enable();
+    if (!locals.includes(this)) {
+      locals.push(this);
+    }
+    let resource = getCurrentResource();
+    if (!resource) {
+      // There is no resource, we need to create a fake one
+      resource = {};
+      idResourcePairing.set(executionAsyncId(), new WeakReference(resource));
+      process.nextTick(() => {
+        resource;
+      });
+      // This prevents the fake resource from being garbage collected too soon
+    }
+    const store = new Map();
+    resource[this.kResourceStore] = store;
+    return store;
+  }
+
+  _exit() {
+    const resource = getCurrentResource();
+    if (resource) {
+      delete resource[this.kResourceStore];
+    }
+  }
+
+  runSyncAndReturn(callback, ...args) {
+    this._enter();
+    try {
+      return callback(...args);
+    } finally {
+      this._exit();
+    }
+  }
+
+  exitSyncAndReturn(callback, ...args) {
+    const resource = getCurrentResource();
+    if (!resource) {
+      return callback(...args);
+    }
+    const store = resource[this.kResourceStore];
+    try {
+      return callback(...args);
+    } finally {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const resource = getCurrentResource();
+    if (resource) {
+      return resource[this.kResourceStore];
+    }
+  }
+
+  /**
+   * after this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  run(callback, ...args) {
+    this._enter();
+    process.nextTick(() => callback(...args));
+    this._exit();
+  }
+
+  exit(callback, ...args) {
+    const resource = getCurrentResource();
+    if (!resource) {
+      return callback(...args);
+    }
+    const store = resource[this.kResourceStore];
+    delete resource[this.kResourceStore];
```

Comment:
As I mentioned earlier, I don't see any value in such `.exit(cb)` method. But in any case, this line of code doesn't seem right. It has to be scoped, yet it messes up the store in current resource synchronously, so subsequent `.getStore()` will return `underfined`. Also, it's strange that tests don't cover this weird situation.